lineoperations/src/linefunctions.c

```diff
@@ -0,0 +1,328 @@
+				{
+					to_remove[i] = FALSE;
+					to_remove[j] = FALSE;
+				}
+
+	/* copy **lines into 'new_file' if it is not FALSE(not duplicate) */
+	for(i = 0; i < total_num_lines; i++)
+		if(!to_remove[i])
+			nf_end = g_stpcpy(nf_end, lines[i]);
+
+    /* set new document */
+	sci_set_text(doc->editor->sci, new_file);
+
+	/* free used memory */
+	for(i = 0; i < total_num_lines; i++)
+		g_free(lines[i]);
+	g_free(lines);
+	g_free(new_file);
+	g_free(to_remove);
+}
+
+
+/* Remove Empty Lines */
+void rmemtyln(GeanyDocument *doc) {
+	gint  total_num_lines;    /* number of lines in the document */
+	gint  i;                  /* iterator */
+
+	total_num_lines = sci_get_line_count(doc->editor->sci);
+
+	sci_start_undo_action(doc->editor->sci);
+
+	for(i = 0; i < total_num_lines; i++)    /* loop through opened doc */
+	{
+		if(sci_get_position_from_line(doc->editor->sci, i) ==
+		   sci_get_line_end_position(doc->editor->sci, i))
+		{
+			scintilla_send_message(doc->editor->sci,
+                                   SCI_DELETERANGE,
+                                   sci_get_position_from_line(doc->editor->sci, i),
+                                   sci_get_line_length(doc->editor->sci, i));
+			total_num_lines--;
+			i--;
+		}
+	}
+
+	sci_end_undo_action(doc->editor->sci);
+}
+
+
+
+
+/* Remove Whitespace Lines */
+void rmwhspln(GeanyDocument *doc) {
+	gint total_num_lines;  /* number of lines in the document */
+	gint indent;
+	gint i;                /* iterator */
+
+	total_num_lines = sci_get_line_count(doc->editor->sci);
+	
+	sci_start_undo_action(doc->editor->sci);
+
+	for(i = 0; i < total_num_lines; i++)    /* loop through opened doc */
+	{
+		indent = scintilla_send_message(doc->editor->sci,
+                                    SCI_GETLINEINDENTPOSITION,
+                                    i, 0);
+
+		if(indent -
+           sci_get_position_from_line(doc->editor->sci, i) ==
+           sci_get_line_end_position(doc->editor->sci, i)-
+           sci_get_position_from_line(doc->editor->sci, i))
+		{
+			scintilla_send_message(doc->editor->sci,
+                                   SCI_DELETERANGE,
+                                   sci_get_position_from_line(doc->editor->sci, i),
+                                   sci_get_line_length(doc->editor->sci, i));
+			total_num_lines--;
+			i--;
+		}
+
+	}
+
+
+	sci_end_undo_action(doc->editor->sci);
+}
+
+
+/* Sort Lines Ascending and Descending */
+void sortlines(GeanyDocument *doc, gboolean asc) {
+	gint  total_num_chars;  /* number of characters in the document */
+	gint  total_num_lines;  /* number of lines in the document */
+	gchar **lines;          /* array to hold all lines in the document */
+	gchar *new_file;        /* *final* string to replace current document */
+	gint  i;                /* iterator */
+
+	total_num_chars = sci_get_length(doc->editor->sci);
+	total_num_lines = sci_get_line_count(doc->editor->sci);
+	lines           = g_malloc(sizeof(gchar *) * total_num_lines+1);
```

Comment:
there's a subtle bug here: the +1 is not applied correctly as `*` has priority over `+`.  Wrap in parentheses. (thanks to Valgrind for finding this)